m4/ax_lua.m4

```diff
@@ -205,10 +210,36 @@ AC_DEFUN([AX_PROG_LUA],
   dnl Make LUA a precious variable.
   AC_ARG_VAR([LUA], [The Lua interpreter, e.g. /usr/bin/lua5.1])
 
+  dnl Figure out whether we should expose LuaJIT as a user facing configure
+  dnl flag and if so whether the default option should be with or without.
+  AS_CASE(["m4_default([$5], [never])"],
```

Comment:
Probably this should be an m4_case, not an AS_CASE, right?

See e.g. m4/ax_is_release.m4:
```m4
AC_DEFUN([AX_IS_RELEASE],[
    AC_BEFORE([AC_INIT],[$0])

    m4_case([$1],
      [git-directory],[
        # $is_release = (.git directory does not exist)
        AS_IF([test -d ${srcdir}/.git || (test -f ${srcdir}/.git && grep \.git/worktrees ${srcdir}/.git)],[ax_is_release=no],[ax_is_release=yes])
      ],
      [minor-version],[
        # $is_release = ($minor_version is even)
        minor_version=`echo "$PACKAGE_VERSION" | sed 's/[[^.]][[^.]]*.\([[^.]][[^.]]*\).*/\1/'`
        AS_IF([test "$(( $minor_version % 2 ))" -ne 0],
              [ax_is_release=no],[ax_is_release=yes])
      ],
      [micro-version],[
        # $is_release = ($micro_version is even)
        micro_version=`echo "$PACKAGE_VERSION" | sed 's/[[^.]]*\.[[^.]]*\.\([[^.]]*\).*/\1/'`
        AS_IF([test "$(( $micro_version % 2 ))" -ne 0],
              [ax_is_release=no],[ax_is_release=yes])
      ],
      [dash-version],[
        # $is_release = ($PACKAGE_VERSION has a dash)
        AS_CASE([$PACKAGE_VERSION],
                [*-*], [ax_is_release=no],
                [*], [ax_is_release=yes])
      ],
      [always],[ax_is_release=yes],
      [never],[ax_is_release=no],
      [
        m4_if($1, [], m4_fatal([$0: invalid policy $1. Valid policies: git-directory, minor-version, micro-version, dash-version, always, never]))
      ])
])
```

Instead of expanding a shell:
```sh
case never in
    never)
        default_luajit=no
        with_luajit=no
        ;;
    allow)
        defaut_luajit=no
        # m4_if([$5], [allow], ...) returned false here
        test "x$with_luajit" != 'xyes' && with_luajit=no
        ;;
    prefer)
        .........
    *)
        as_fn_error $? "Unrecognized value for ENABLE_LUAJIT"
        ;;
esac
```

it just emits only the code for the branch you want.

Reply:
That seems reasonable, and saves the extra step of repeating the check with `m4_if` just to keep the flag help from being hoisted out twice.

Reply:
How does e801fec look, it seems to give the results I expect.

Reply:
Any particular reason you didn't opt to use the pattern I mentioned above:

```m4
      [
        m4_if($1, [], m4_fatal([$0: invalid policy $1. Valid policies: git-directory, minor-version, micro-version, dash-version, always, never]))
      ])
```
and instead used:
```m4
    [
      AC_MSG_ERROR([Unrecognized value for ENABLE_LUAJIT])
    ]
```

Using AC_MSG_ERROR means that code is successfully generated into ./configure which unconditionally errors out -- whereas using `m4_fatal` would mean that `autoreconf -fi` can fail immediately if you pass the wrong value to `AX_PROG_LUA`. See e.g. 2dde1ac11b284e9e75cb78082e81592926c6d7cd

Also IMHO it's slightly cleaner than the `AM_CONDITIONAL([LUAJIT], [test "x$with_luajit" == "xyes"])` if you can avoid generating code at all. I wasn't sure how exactly to do this without inlining `AM_COND_IF([LUAJIT],` into each branch. It may be possible to do this by defining it as a private macro e.g. `_LUA_FINDINTERP` and calling it in each branch of m4_case, not sure if it's worth it.

Reply:
> Any particular reason [...]

Because that was a full blown example from another macro and we were chasing a different issue at that point I didn't recognize that as a relevant difference with it's own advantages.

Thanks for the explanation, and yes failing at autoconf time is better. I just implemented that pattern.

As for trying to avoid the later use of `AM_CONDITIONAL` I can see ways to rig it up inside of this macro (although the result would be quite a bit more convoluted and I don't see any significant gains to make it that complex), but more than that it is very useful to be able to use `AM_CONF_IF` *outside* of this macro later. So that seems like all-pain and no-gain to me.